app/src/main/java/com/example/playlistmakervmp/SearchActivity.kt

```diff
@@ -177,6 +228,66 @@ class SearchActivity : AppCompatActivity() {
             placeholderNothingFoundText.visibility = View.GONE
             searchTracks()
         }
+
+        trackAdapter.setOnTrackClickListener(object : OnTrackClickListener {
+            override fun onTrackClick(position: Int) {
+                val editor = sharedPrefs.edit()
+                if ((history.size <= 9) and !isTrackInHistory(tracks[position])) {
+                    history.add(0, tracks[position])
+                    json = Gson().toJsonTree(history).asJsonArray.toString()
+                    editor.putString("TRACKS", json).apply()
+                }
+                else if ((history.size == 10) and !isTrackInHistory(tracks[position])) {
+
+                    history.add(0, tracks[position])
+                    history.removeLast()
+                    json = Gson().toJsonTree(history).asJsonArray.toString()
+                    editor.putString("TRACKS", json).apply()
+                }
+                else if (isTrackInHistory(tracks[position])) {
+
+                    lateinit var songbuf: Track
+                    for (song in history) {
+                        if (song.trackId == tracks[position].trackId) {
+                            songbuf = song
+                        }
+                    }
+                    history.remove(songbuf)
+                    history.add(0, tracks[position])
+                    json = Gson().toJsonTree(history).asJsonArray.toString()
+                    editor.putString("TRACKS", json).apply()
+                }
+            }
+        })
+        editText.setOnFocusChangeListener { view, hasFocus ->
+            if (history.size > 0) {
+
+                searched.visibility = if (hasFocus and editText.text.isEmpty()) View.VISIBLE else View.GONE
+                clearHistory.visibility =
+                    if (hasFocus and editText.text.isEmpty()) View.VISIBLE else View.GONE
+                recyclerView.visibility = if (hasFocus and editText.text.isEmpty()) View.VISIBLE else View.GONE
+                json = sharedPrefs.getString("TRACKS", "")
+                history = Gson().fromJson(json, listType)
+                recyclerView.adapter = historyAdapter
+                historyAdapter.notifyDataSetChanged()
+
+            }
+        }
+        clearHistory.setOnClickListener {
+            history.clear()
+            historyAdapter.notifyDataSetChanged()
+            sharedPrefs.edit().putString("TRACKS", "").apply()
+            searched.visibility = View.GONE
+            clearHistory.visibility = View.GONE
+            recyclerView.visibility = View.GONE
+        }
+
+    }
+    fun isTrackInHistory(track: Track): Boolean {
+        for (song in history) {
+            if (track.trackId == song.trackId) return true
+        }
+        return false
```

Comment:
```suggestion
        return history.any { it.trackId == track.trackId }
```
В котлине очень много полезных функций для работы со списками, лучше использовать их, чтобы не писать свои.